Make trivial thread-idx for degenerate statements without thread-idx.

zheng-xq opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #36563 Add sync-point insertions and block/thread local memory allocations
* **#36480 Make trivial thread-idx for degenerate statements without thread-idx.**

Differential Revision: [D20992505](https://our.internmc.facebook.com/intern/diff/D20992505)

zheng-xq committed:
Make trivial thread-idx for degenerate statements without thread-idx.

[ghstack-poisoned]

zheng-xq committed:
Update on "Make trivial thread-idx for degenerate statements without thread-idx."


Differential Revision: [D20992505](https://our.internmc.facebook.com/intern/diff/D20992505)

[ghstack-poisoned]

zheng-xq committed:
Update on "Make trivial thread-idx for degenerate statements without thread-idx."


Differential Revision: [D20992505](https://our.internmc.facebook.com/intern/diff/D20992505)

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/tensorexpr/cuda_codegen.cpp:
```diff
@@ -469,6 +475,132 @@ std::string CudaCodeGen::GetUniqueFuncName(const std::string& func_prefix) {
+ private:
+  Stmt* rewrite(const std::vector<Stmt*>& stmts) {
+    std::vector<Stmt*> cloned_stmts(stmts.size());
+    for (int index = 0; index < stmts.size(); index++) {
```
Please fix the clang-tidy complaint :)

zheng-xq replied:
Done

ZolotukhinM commented on torch/csrc/jit/tensorexpr/cuda_codegen.cpp:
```diff
@@ -469,6 +475,132 @@ std::string CudaCodeGen::GetUniqueFuncName(const std::string& func_prefix) {
 }
 
+// Find all the statements that are not covered by any thread-idx axes,
+// and wrapp them under a trivial thread idx.
```
Typo: wrapp.

zheng-xq replied:
Done

ZolotukhinM commented on torch/csrc/jit/tensorexpr/cuda_codegen.cpp:
```diff
@@ -469,6 +475,132 @@ std::string CudaCodeGen::GetUniqueFuncName(const std::string& func_prefix) {
+    }
+    Stmt* new_block = Block::make(cloned_stmts);
+    // Wrap the new block under a trivial thread-idx
+    //   for t in 0..1: // threadIdx
+    //     if (t < 1):
+    //       new_block
```
Not a concern for this PR, but this structure seems like an overkill for just setting a thread-idx for new_block. Is there another way we can represent it?

zheng-xq replied:
Add a note that this serves too purposes: 1. setting the mask; 2. setting the sync-point boundary.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/cuda_codegen.cpp:
```diff
@@ -469,6 +475,132 @@ std::string CudaCodeGen::GetUniqueFuncName(const std::string& func_prefix) {
+    std::list<Stmt*> old_stmts = v->stmts();
+    std::vector<bool> need_rewrites(old_stmts.size());
+    std::vector<Stmt*> new_stmts(old_stmts.size());
+    ;
```
Nit: unnecessary ";"

zheng-xq replied:
done

ZolotukhinM commented on torch/csrc/jit/tensorexpr/cuda_codegen.cpp:
```diff
@@ -469,6 +475,132 @@ std::string CudaCodeGen::GetUniqueFuncName(const std::string& func_prefix) {
+    if (all_need_fix) {
+      Stmt* parent = v->get_parent();
+      For* loop_parent = dynamic_cast<For*>(parent);
+      if (loop_parent) {
```
Nit: can do "`if (loop_parent && loop_parent->loop_options().is_gpu_block_index()) {`

ZolotukhinM commented on torch/csrc/jit/tensorexpr/cuda_codegen.cpp:
```diff
@@ -469,6 +475,132 @@ std::string CudaCodeGen::GetUniqueFuncName(const std::string& func_prefix) {
+      }
+
+      // Rewrite the stmts from [start, stop)
+      std::vector<Stmt*> stmts_block(
```
Nit: the name `stmts_block` is slightly misleading as it's not a `Block` (even though it's very similar by its meaning).

ZolotukhinM commented on torch/csrc/jit/tensorexpr/cuda_codegen.cpp:
```diff
@@ -469,6 +475,132 @@ std::string CudaCodeGen::GetUniqueFuncName(const std::string& func_prefix) {
+
+      start = stop;
+    }
+    Stmt* rewritten_block = Block::make(rewrite_stmts);
```
Maybe add an assert that rewrite_stmts.size() == new_stmts.size() == old_stmts.size()?

ZolotukhinM commented on torch/csrc/jit/tensorexpr/cuda_codegen.cpp:
```diff
@@ -469,6 +475,132 @@ std::string CudaCodeGen::GetUniqueFuncName(const std::string& func_prefix) {
+    }
+
+    // If all needs fix, check if parent is a GPU block
+    if (all_need_fix) {
```
Why do we need to special-case this? Won't the logic below handle it as well?

zheng-xq replied:
Added a note to clarify.

ZolotukhinM reviewed (approved):
Looks good overall, some comments are inline!

nickgg commented on test/cpp/tensorexpr/test_cuda.cpp:
```diff
@@ -477,6 +477,103 @@ void testCudaOneBlockMultiThreadGlobalReduce1() {
+    b_ref(i) = i;
+  }
+
+  float* a_dev = nullptr;
```
Running the codegen is not sufficient here, I think. Can you verify the correct code was generated in a literal way?

zheng-xq replied:
Added a TODO to add the check in a follow-up.

zheng-xq committed:
Update on "Make trivial thread-idx for degenerate statements without thread-idx."


Differential Revision: [D20992505](https://our.internmc.facebook.com/intern/diff/D20992505)

[ghstack-poisoned]

zheng-xq committed:
Update on "Make trivial thread-idx for degenerate statements without thread-idx."


Differential Revision: [D20992505](https://our.internmc.facebook.com/intern/diff/D20992505)

[ghstack-poisoned]
